libsolidity/ast/Types.cpp

```diff
@@ -571,39 +571,104 @@ TypePointer FixedPointType::binaryOperatorResult(Token::Value _operator, TypePoi
 	return commonType;
 }
 
-tuple<bool, rational> RationalNumberType::isValidLiteral(Literal const& _literal)
+tuple<bool, rational> RationalNumberType::parseRational(string const& _value)
 {
 	rational x;
 	try
 	{
-		rational numerator;
-		rational denominator(1);
-		
-		auto radixPoint = find(_literal.value().begin(), _literal.value().end(), '.');
-		if (radixPoint != _literal.value().end())
+		auto radixPoint = find(_value().begin(), _value().end(), '.');
+
+		if (radixPoint != _value().end())
 		{
+			if (!_allowRadixPoint)
+				return make_tuple(false, rational(0));
+
 			if (
-				!all_of(radixPoint + 1, _literal.value().end(), ::isdigit) || 
-				!all_of(_literal.value().begin(), radixPoint, ::isdigit) 
+				!all_of(radixPoint + 1, _value().end(), ::isdigit) ||
+				!all_of(_value().begin(), radixPoint, ::isdigit)
 			)
 				return make_tuple(false, rational(0));
-			//Only decimal notation allowed here, leading zeros would switch to octal.
+
+			// Only decimal notation allowed here, leading zeros would switch to octal.
 			auto fractionalBegin = find_if_not(
-				radixPoint + 1, 
-				_literal.value().end(), 
+				radixPoint + 1,
+				_value().end(),
 				[](char const& a) { return a == '0'; }
 			);
 
-			denominator = bigint(string(fractionalBegin, _literal.value().end()));
+			rational numerator;
+			rational denominator(1);
+
+			denominator = bigint(string(fractionalBegin, _value().end()));
 			denominator /= boost::multiprecision::pow(
-				bigint(10), 
-				distance(radixPoint + 1, _literal.value().end())
+				bigint(10),
+				distance(radixPoint + 1, _value().end())
 			);
-			numerator = bigint(string(_literal.value().begin(), radixPoint));
+			numerator = bigint(string(_value().begin(), radixPoint));
 			x = numerator + denominator;
 		}
 		else
+			x = bigint(_value());
+		return make_tuple(true, x);
+	}
+	catch (...)
+	{
+		return make_tuple(false, rational(0));
+	}
+}
+
+tuple<bool, rational> RationalNumberType::isValidLiteral(Literal const& _literal)
+{
+	rational x;
+	try
+	{
+		auto expPoint = _literal.value().find_first_of("eE");
+
+		if (starts_with(_literal.value.begin(), _literal.value().end(), "0x"))
```

Comment:
Where does `starts_with` come from? Can you use `starts_with(_literal.value, "0x")`?